ReactiveCocoa/Swift/Signal.swift

```diff
@@ -505,26 +505,42 @@ public func skipWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signa
 /// instead, regardless of whether `signal` has sent events already.
 public func takeUntilReplacement<T, E>(replacement: Signal<T, E>)(signal: Signal<T, E>) -> Signal<T, E> {
 	return Signal { observer in
-		let signalDisposable = SerialDisposable()
-
-		let replacementDisposable = replacement.observe(next: { value in
-			signalDisposable.dispose()
-			sendNext(observer, value)
-		}, error: { error in
-			sendError(observer, error)
-		}, completed: {
-			sendCompleted(observer)
-		})
-
-		if !signalDisposable.disposed {
-			signalDisposable.innerDisposable = signal.observe(next: { value in
-				sendNext(observer, value)
-			}, error: { error in
-				sendError(observer, error)
+		/// When this disposable is disposed, we'll stop forwarding events from `signal` to
+		/// the observer. It's atomic because we need to defend against the case that the
+		/// replacement signal sends an event while we're starting to observe `signal`.
+		let signalDisposableAtomic = Atomic(SerialDisposable())
```

Comment:
Hmm… it's a strange composition. I feel like the `SerialDisposable` should be enough here, since it has an internal `Atomic` and is intended to make things related to disposal happen in serial. But the internal `Atomic` holds private implementation, so that's a dead end.